lib/reversetunnel/remotesite.go

```diff
@@ -151,16 +222,112 @@ func (s *remoteSite) GetName() string {
 }
 
 func (s *remoteSite) GetLastConnected() time.Time {
-	s.Lock()
-	defer s.Unlock()
-	return s.lastActive
+	connInfo, err := s.getLatestTunnelConnection()
+	if err != nil {
+		return time.Time{}
+	}
+	return connInfo.GetLastHeartbeat()
+}
+
+func (s *remoteSite) periodicSendDiscoveryRequests() {
+	ticker := time.NewTicker(defaults.ReverseTunnelAgentHeartbeatPeriod)
+	defer ticker.Stop()
+	if err := s.sendDiscoveryRequest(); err != nil {
+		s.Warningf("failed to send discovery: %v", err)
+	}
+	for {
+		select {
+		case <-s.ctx.Done():
+			s.Debugf("closing")
+			return
+		case <-ticker.C:
+			err := s.sendDiscoveryRequest()
+			if err != nil {
+				s.Warningf("could not send discovery request: %v", trace.DebugReport(err))
+			}
+		}
+	}
+}
+
+func (s *remoteSite) isOnline(conn services.TunnelConnection) bool {
+	diff := s.clock.Now().Sub(conn.GetLastHeartbeat())
+	return diff < defaults.ReverseTunnelOfflineThreshold
+}
+
+// findDisconnectedProxies
+func (s *remoteSite) findDisconnectedProxies() ([]services.Server, error) {
+	conns, err := s.srv.AccessPoint.GetTunnelConnections(s.domainName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	connected := make(map[string]bool)
+	for _, conn := range conns {
+		if s.isOnline(conn) {
+			connected[conn.GetProxyName()] = true
+		}
+	}
+	proxies, err := s.srv.AccessPoint.GetProxies()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var missing []services.Server
+	for i := range proxies {
+		proxy := proxies[i]
+		if !connected[proxy.GetName()] {
+			missing = append(missing, proxy)
+		}
+	}
+	return missing, nil
+}
+
+func (s *remoteSite) sendDiscoveryRequest() error {
```

Comment:
Comment needed for non-trivial function.

Reply:
:white_check_mark: